File: haskell-fuel-engine/src/FuelUpload/Report.hs
module FuelUpload.Report
  ( OperationalBatchStatus (..)
  , OperationalQuarantinedRow (..)
  , OperationalBatchReport (..)
  , projectOperationalReport
  ) where

import FuelUpload.Domain.Decision
import FuelUpload.Domain.Primitive
import FuelUpload.Domain.Row

data OperationalBatchStatus
  = OperationalReady
  | OperationalFatal
  deriving stock (Eq, Show)

data OperationalQuarantinedRow = OperationalQuarantinedRow
  { operationalQuarantinedRowNumber :: RowNumber
  , operationalQuarantineReasons :: [QuarantineReason]
  }
  deriving stock (Eq, Show)

data OperationalBatchReport = OperationalBatchReport
  { operationalStatus :: OperationalBatchStatus
  , operationalCounts :: BatchSummary
  , operationalUploadedTransactionIds :: [TransactionId]
  , operationalRejectedRowNumbers :: [RowNumber]
  , operationalQuarantinedRows :: [OperationalQuarantinedRow]
  , operationalSkippedDuplicateRowNumbers :: [RowNumber]
  , operationalFatalErrors :: [FatalError]
  }
  deriving stock (Eq, Show)

projectOperationalReport :: BatchDecision -> OperationalBatchReport
projectOperationalReport decision =
  OperationalBatchReport
    { operationalStatus = status
    , operationalCounts = batchSummary decision
    , operationalUploadedTransactionIds =
        case status of
          OperationalReady -> acceptedTransactionIds (batchRows decision)
          OperationalFatal -> []
    , operationalRejectedRowNumbers = rejectedRowNumbers (batchRows decision)
    , operationalQuarantinedRows = quarantinedRows (batchRows decision)
    , operationalSkippedDuplicateRowNumbers = skippedDuplicateRowNumbers (batchRows decision)
    , operationalFatalErrors = fatalErrors
    }
  where
    (status, fatalErrors) =
      case batchOutcome decision of
        BatchUploadable -> (OperationalReady, [])
        BatchBlockedByFatal errors -> (OperationalFatal, nonEmptyToList errors)

acceptedTransactionIds :: [RowDecision] -> [TransactionId]
acceptedTransactionIds =
  foldr collect []
  where
    collect row ids =
      case row of
        Accepted transaction -> transactionId transaction : ids
        AcceptedWithWarnings transaction _ -> transactionId transaction : ids
        Quarantined _ _ -> ids
        SkippedDuplicate _ -> ids
        Rejected _ -> ids
        Fatal _ -> ids

rejectedRowNumbers :: [RowDecision] -> [RowNumber]
rejectedRowNumbers =
  foldr collect []
  where
    collect row numbers =
      case row of
        Rejected rejected -> parsedRowNumber (rejectedRow rejected) : numbers
        Accepted _ -> numbers
        AcceptedWithWarnings _ _ -> numbers
        Quarantined _ _ -> numbers
        SkippedDuplicate _ -> numbers
        Fatal _ -> numbers

quarantinedRows :: [RowDecision] -> [OperationalQuarantinedRow]
quarantinedRows =
  foldr collect []
  where
    collect row quarantines =
      case row of
        Quarantined transaction reasons ->
          OperationalQuarantinedRow
            { operationalQuarantinedRowNumber = transactionRowNumber transaction
            , operationalQuarantineReasons = nonEmptyToList reasons
            }
            : quarantines
        Accepted _ -> quarantines
        AcceptedWithWarnings _ _ -> quarantines
        SkippedDuplicate _ -> quarantines
        Rejected _ -> quarantines
        Fatal _ -> quarantines

skippedDuplicateRowNumbers :: [RowDecision] -> [RowNumber]
skippedDuplicateRowNumbers =
  foldr collect []
  where
    collect row numbers =
      case row of
        SkippedDuplicate skipped -> parsedRowNumber (skippedRow skipped) : numbers
        Accepted _ -> numbers
        AcceptedWithWarnings _ _ -> numbers
        Quarantined _ _ -> numbers
        Rejected _ -> numbers
        Fatal _ -> numbers

nonEmptyToList :: Foldable f => f a -> [a]
nonEmptyToList = foldr (:) []

File: rust-fuel-engine/src/domain/report.rs
use super::Summary;
use super::decision::{BatchDecision, RowDecision};
use super::primitives::{RowNumber, TransactionId};
use super::validation::{FatalError, QuarantineReason};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum OperationalBatchStatus {
    Ready,
    Fatal,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct OperationalQuarantinedRow {
    pub row_number: RowNumber,
    pub reasons: Vec<QuarantineReason>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct OperationalBatchReport {
    pub status: OperationalBatchStatus,
    pub counts: Summary,
    pub uploaded_transaction_ids: Vec<TransactionId>,
    pub rejected_row_numbers: Vec<RowNumber>,
    pub quarantined_rows: Vec<OperationalQuarantinedRow>,
    pub skipped_duplicate_row_numbers: Vec<RowNumber>,
    pub fatal_errors: Vec<FatalError>,
}

pub fn project_operational_report(decision: &BatchDecision) -> OperationalBatchReport {
    let status = match decision {
        BatchDecision::Ready { .. } => OperationalBatchStatus::Ready,
        BatchDecision::Blocked { .. } => OperationalBatchStatus::Fatal,
    };

    let uploaded_transaction_ids = if status == OperationalBatchStatus::Fatal {
        Vec::new()
    } else {
        decision
            .rows()
            .iter()
            .filter_map(|row| match row {
                RowDecision::Accepted(transaction) | RowDecision::Warning { transaction, .. } => {
                    Some(transaction.transaction_id.clone())
                }
                RowDecision::Quarantined { .. }
                | RowDecision::SkippedDuplicate(_)
                | RowDecision::Rejected(_)
                | RowDecision::Fatal(_) => None,
            })
            .collect()
    };

    OperationalBatchReport {
        status,
        counts: decision.summary().clone(),
        uploaded_transaction_ids,
        rejected_row_numbers: decision
            .rows()
            .iter()
            .filter_map(|row| match row {
                RowDecision::Rejected(rejected) => Some(rejected.row_number),
                _ => None,
            })
            .collect(),
        quarantined_rows: decision
            .rows()
            .iter()
            .filter_map(|row| match row {
                RowDecision::Quarantined {
                    transaction,
                    reasons,
                    ..
                } => Some(OperationalQuarantinedRow {
                    row_number: transaction.row_number,
                    reasons: reasons.as_slice().to_vec(),
                }),
                _ => None,
            })
            .collect(),
        skipped_duplicate_row_numbers: decision
            .rows()
            .iter()
            .filter_map(|row| match row {
                RowDecision::SkippedDuplicate(skipped) => Some(skipped.row_number),
                _ => None,
            })
            .collect(),
        fatal_errors: match decision {
            BatchDecision::Ready { .. } => Vec::new(),
            BatchDecision::Blocked { fatal_errors, .. } => fatal_errors.clone(),
        },
    }
}
